### DISCUSSION TITLE
Dynamic Paramethers to definePageMeta

### DISCUSSION THREAD

**Discussion by aligzl** [nuxt/nuxt#17193] [Questions]
Hello,

I would like to add dynamic meta to definePageMeta depends on route params but i got errors like Reproduction below
`/* [nuxt] [request error] route is not defined */`
[Reproduction](https://stackblitz.com/edit/github-xpdkva-qsdoh3?file=pages%2Fitems%2F[id].vue)

```
<script setup>
const route = useRoute();
useHead({
  title: route.params.id == "Add" ? 'Add Product' : 'Edit Product' // Works
});

/*  */
definePageMeta({
  permission: route.params.id == "Add" ? 'Add Product' : 'Edit Product' // Not works here
});
</script>
```

**Comment by ShaggyTech** [comment]
[https://v3.nuxtjs.org/guide/directory-structure/pages#page-metadata](https://v3.nuxtjs.org/guide/directory-structure/pages#page-metadata)

According to the docs:

> Much like defineEmits or defineProps (see [Vue docs ](https://vuejs.org/api/sfc-script-setup.html#defineprops-defineemits)), definePageMeta is a compiler macro. It will be compiled away so you cannot reference it within your component. Instead, the metadata passed to it will be hoisted out of the component. Therefore, the page meta object cannot reference the component (or values defined on the component). However, it can reference imported bindings.

This doesn't really answer your question but explains why it's happening at least.

**Comment by ShaggyTech** [comment]
I'm not 100% clear on what you're trying to use the meta for but I played around with it and came up with this that might work for you.

```ts
useHead({
  title: route.params.id,
  meta: [{ name: 'permission', content: route.params.id }],
});
```

**Reply by aligzl** [reply]
I am using definePageMeta to manage permission for each page by middleware globaly and dynamicly.

Here is my global routemiddleware,
```
export default defineNuxtPlugin(() => {

  const auth = useUser()

  addRouteMiddleware('auth', (to, from) => {
    
    if( ! auth.value && to.path !== '/login' ){
      return navigateTo('/login')
    }

    if(auth.value && auth.value.rank !== "admin"){
      if(to.meta.permission){
        if(!auth.value.permissions.includes(to.meta.permission)){
          return navigateTo('/access-denied?meta=' + to.meta.permission)
        }
      }
    }

  }, { global: true })
})
```

**Reply by jcjp** [reply]
Have you fixed your issue with using `definePageMeta` for `middleware`?

**Comment by bros4president** [comment]
As the documentation posted by @ShaggyTech explains, you can't access variables defined outside definePageMeta because it's a compiler macro. You do have access to imports, so I think replacing `route.params.id` with `useRoute().params.id` should be enough in that case :)

**Reply by adriaandotcom** [reply]
When using

```js
definePageMeta({
  title: useRoute().params.slug,
});
```

I get this error:

```
[nuxt] [request error] nuxt instance unavailable
  at Module.useNuxtApp (./.nuxt/dist/server/server.mjs:925:13)  
  at Module.useRoute (./.nuxt/dist/server/server.mjs:1803:32)  
  at $id_28114ce3 (./.nuxt/dist/server/server.mjs:3406:32)  
  at async __instantiateModule__ (./.nuxt/dist/server/server.mjs:35875:3)
```

I don't really think `definePageMeta` should be used with variables.

**Comment by kyjus25** [comment]
I have a similar issue, has any more headway been made to this?

I am looking to be able to overwrite the default layout based on a `?layout=<foo>` query param. Much like OP, there is no way to get access to the route data.

I have also tried:
```ts
definePageMeta({ layout: false });
setPageLayout(route.query.theme);
```
as the example on https://nuxt.com/docs/guide/directory-structure/layouts#changing-the-layout-dynamically suggests, but I get `setPageLayout should not be called to change the layout on the server within a component as this will cause hydration errors`

So... it's like I'm either deliberately blocked from being able to do this, or there is a better intended way

It seems like the best approach that would solve all of our issues would be to make `definePageMeta` a function that provides the route:
```ts
definePageMeta((route: RouteLocationNormalizedLoaded) => ({
  title: route.params.slug,
  layout: route.query.theme || 'default',
}));
```

To be clear, Nuxt2 already has similar functionality with the context:
https://v2.nuxt.com/docs/directory-structure/layouts#custom-layout

**Reply by immachakata** [reply]
I'm stuck here as well. Is there any workaround

**Reply by itai-gendler** [reply]
me 2

**Reply by jacobhutchings** [reply]
me 3

**Comment by tinawng** [comment]
I managed to find a workaround that fits my use case, maybe it can help others.
```js
const route = useRoute()
definePageMeta({ theme: "" })
route.meta.theme = route.params.theme
```
I use the `theme` variable for styling, reactivity seems to work fine aswell with this trick.

**Reply by imrim12** [reply]
Nuxt has the useRoute composable typed properly, so you can use it like this for typing

```
const route = useRoute('your-route-name-explicitly')
definePageMeta({ theme: "" })
route.meta.theme = route.params.theme // this is typed
```

**Reply by ASoleymani76** [reply]
This works perfectly
